portfolio/src/main/java/com/google/sps/data/Comment.java

```diff
@@ -14,28 +14,121 @@
+import com.google.cloud.translate.Translation;
+
 /** A comment on my portfolio */
 public final class Comment {
 
-  private final long id;
-  private final String content;
-  private final long time;
-  private long thumbsup;
-  private long thumbsdown;
-  private long popularity;
-  private final String name;
-  private final String email;
-  private final String username;
-
-  public Comment(long id, String content, long time, long thumbsup, long thumbsdown, String name, String email, String username) {
-    this.id = id;
-    this.content = content;
-    this.time = time;
-    this.thumbsup = thumbsup;
-    this.thumbsdown = thumbsdown;
-    this.popularity = thumbsup - thumbsdown;
-    this.name = name;
-    this.email = email;
-    this.username = username;
-  }
+    private final long id;
+    private String content;
+    private final long time;
+    private long thumbsup;
+    private long thumbsdown;
+    private long popularity;
+    private final String name;
+    private final String email;
+    private final String username;
+
+    public Comment(Entity commentEntity) {
+        this.id = (long) commentEntity.getKey().getId();
+        this.content = (String) commentEntity.getProperty("content");
+        this.time = (long) commentEntity.getProperty("time");
+        this.thumbsup = (long) commentEntity.getProperty("thumbsup");
+        this.thumbsdown = (long) commentEntity.getProperty("thumbsdown");
+        this.popularity = (long) commentEntity.getProperty("popularity");
+        this.name = (String) commentEntity.getProperty("name");
+        this.email = (String) commentEntity.getProperty("email");
+        this.username = (String) commentEntity.getProperty("username");
+    }
+
+     public void incrementThumbsup() {
+        this.thumbsup++;
+        return;
+    }
+
+    public void decrementThumbsup() {
+        this.thumbsup--;
+        return;
+    }
+
+    public void incrementThumbsdown() {
+        this.thumbsdown++;
+        return;
+    }
+
+    public void decrementThumbsdown() {
+        this.thumbsdown--;
+        return;
+    }
+
+    public void incrementPopularity() {
+        this.popularity++;
+        return;
+    }
+
+    public void decrementPopularity() {
+        this.popularity--;
+        return;
+    }
+
+    public void translateComment(String language) {
+        Translate translate = TranslateOptions.getDefaultInstance().getService();
+
+        try {
+            // Translate comment
+            Translation translation = translate.translate(this.content, Translate.TranslateOption.targetLanguage(language));
+            this.content = translation.getTranslatedText();
+            return;
+        } catch (RuntimeException e) {
+            return;
+        }
+    }
+
+    public void updateDatabase(Entity commentEntity) {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+        Transaction txn = datastore.beginTransaction();
```

Comment:
The problem here is that the read happens outside of the transaction. You read the comment earlier, save its state and then start a transaction to change some values. The problem is that the comment you read earlier isn't necessarily valid anymore. It might have been updated in the database in between.

So in this case you have:
1. read comment entity
2. begin transaction
3. modify values
4. end transaction

What you actually want is:
1. begin transaction
2. read comment entity
3. update values
4. end transaction

Reply:
Makes sense. I submitted changes to address this. I am now starting a transaction as soon as I need to read the comments entity and the user info entity. I use the same transaction when I am modifying their values and only close the transaction after that has already happened.